domain/secretbackend/state/operation.go

```diff
@@ -0,0 +1,202 @@
+VALUES (
+    $SecretBackend.uuid,
+    $SecretBackend.name,
+    $SecretBackend.backend_type,
+    $SecretBackend.token_rotate_interval
+)
+ON CONFLICT (uuid) DO UPDATE SET
+    name=EXCLUDED.name,
+    token_rotate_interval=EXCLUDED.token_rotate_interval;`, SecretBackend{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.upsertRotationStmt, err = sqlair.Prepare(`
+INSERT INTO secret_backend_rotation
+    (backend_uuid, next_rotation_time)
+VALUES (
+    $SecretBackendRotation.backend_uuid,
+    $SecretBackendRotation.next_rotation_time
+)
+ON CONFLICT (backend_uuid) DO UPDATE SET next_rotation_time=EXCLUDED.next_rotation_time;`,
+		SecretBackendRotation{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.clearConfigStmt, err = sqlair.Prepare(`
+DELETE FROM secret_backend_config
+WHERE backend_uuid = $M.uuid AND name NOT IN ($S[:]);`,
+		sqlair.M{}, sqlair.S{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.upsertConfigStmt, err = sqlair.Prepare(`
+INSERT INTO secret_backend_config
+    (backend_uuid, name, content)
+VALUES (
+    $SecretBackendConfig.backend_uuid,
+    $SecretBackendConfig.name,
+    $SecretBackendConfig.content
+)
+ON CONFLICT (backend_uuid, name) DO UPDATE SET content = EXCLUDED.content`,
+		SecretBackendConfig{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (o *upsertOperation) validate(ctx context.Context, tx *sqlair.TX) error {
+	var existing SecretBackend
+	err := tx.Query(ctx, o.getBackendStmt, sqlair.M{"uuid": o.ID}).Get(&existing)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		// New insert.
+		if o.Name == "" {
+			return fmt.Errorf("backend name is missing: %w", backenderrors.NotValid)
+		}
+		if o.BackendType == "" {
+			return fmt.Errorf("backend type is missing: %w", backenderrors.NotValid)
+		}
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+	// Update.
+	if existing.BackendType == "" {
+		// This should never happen, but just in case.
+		return errors.Errorf("backend type is empty for backend %q", o.ID)
+	}
+	if existing.Name == "" {
+		// This should never happen, but just in case.
+		return errors.Errorf("backend name is empty for backend %q", o.ID)
+	}
+	if o.BackendType != "" && o.BackendType != existing.BackendType {
+		// The secret backend type is immutable.
+		return errors.Errorf(
+			"cannot change backend type from %q to %q because backend type is immutable",
+			existing.BackendType, o.BackendType,
+		)
+	}
+	// Fill in the existing backend type.
+	o.BackendType = existing.BackendType
+	if o.Name == "" {
+		// Fill in the existing name.
+		o.Name = existing.Name
+	}
+	if o.TokenRotateInterval == nil && existing.TokenRotateInterval.Valid {
+		// Fill in the existing token rotate interval.
+		o.TokenRotateInterval = &existing.TokenRotateInterval.Duration
+	}
+	return nil
+}
+
+// Apply applies the upsert operation to the database.
+func (o *upsertOperation) Apply(ctx context.Context, tx *sqlair.TX) error {
```

Comment:
Do Apply() and Prepare() need to be exported? They are only used in this package